tiledb/sm/misc/parallel_functions.h

```diff
@@ -270,17 +221,7 @@ Status parallel_for(
   }
 
   // Wait for all instances of `execute_subrange` to complete.
-  // This is ignoring the wait status as we use failed_exception for propagating
-  // the tasks exceptions.
-  std::ignore = tp->wait_all(tasks);
-
-  if (failed_exception.has_value()) {
-    std::rethrow_exception(failed_exception.value());
-  }
-  if (failed_status.has_value()) {
-    return failed_status.value();
-  }
-  return Status{};  // otherwise return OK
+  return tp->wait_all(tasks);
```

Comment:
It's a regression to go back to returning `Status`. That cannot stand.

It's how to throw exceptions that's at issue. It's not as simple as it might seem.
* The existing code mimics its prior behavior, which was to do its own exception handling, regardless of how the thread pool worked. There's some value in this approach because it decouples thread pool behavior from `parallel_for` behavior, allowing the thread pool to evolve separately.
  * It's not clear that's a benefit we actually need.
  * We don't have any documented policy about conjunctive failure (all segments must execute) and disjunctive failure (segments might not execute if there's an exception somewhere). Right now we have undocumented conjunctive behavior, but that does not perform well in failure cases. We have the possibility of adding a policy template argument to `parallel_for` to allow both cases. We can implement such a behavior difference entirely here, without needing to change the thread pool, or we could do it by adding a related policy to the relevant thread pool function.
* The present code relies on the thread pool to deal with failure. The relevant thread pool function doesn't throw, but only returns a single status. Changing the return statement here to use `throw_if_not_ok()` is the simplest way to get `parallel_for` to throw.
  * If that's what we want to do, then at the least we need to document that the error behavior of `parallel_for` is based on that of the thread pool. That behavior needs to be documented, so this PR would need documentation added both here and at the thread pool, which is not explicit on this point.

Reply:
The proposed behavior in the P2300 C++ proposal is to throw if there is an error in any executing thread (what is done in the case of multiple failures isn't completely specified).   There are currently six calls to `wait_all` and two calls to `wait_all_status` in our code (outside of the unit test for `ThreadPool`).  Having `ThreadPool` throw a compound exception would be a straightforward thing to do in `ThreadPool` and a straightforward thing to propagate.  `parallel_for` could then do a try/catch on `wait_all` -- though `throw_if_not_ok` is a simpler solution if we want the caller of `parallel_for` rather than `parallel_for` itself to handle the exception (which could also be rethrown).

Since `ThreadPool` uses `async` to create futures for each submitted task, it has to handle any exceptions thrown by the tasks, since `future::get` will throw if there was an exception in the corresponding task. It has to be conjunctive in some sense.  We could, conceivably, cancel any tasks that haven't yet launched, but this would be really messy as `wait_all` only has a vector of futures and there isn't any way of directly getting (much less cancelling) a task just given a future.  We still could have multiple exceptions even if we were able to cancel tasks before they were run.  This could nonetheless provide some efficiency gains in some cases if there were more tasks than threads to execute them.  However, `parallel_for` partitions its work so that it submits as many tasks as there are threads.  If this resulted in some of the tasks being queued (because the thread pool was doing other work) then some of the `parallel_for` tasks could conceivably be cancelled.  On the other hand, we are (hopefully) striving not to oversubscribe the thread pool.